### PR TITLE
MAINT/STY: use Black formatting

### PR DESCRIPTION
**Pull request by tupui**
I propose to apply [Black](https://black.readthedocs.io) on our code base. This was discussed during the last [community meetup](https://hackmd.io/@tupui/scipy-meetup-3). Using Black would remove all discussions about code style in PRs. It would be the entire responsibility of Black to format code, hence make our code base very consistent and remove the extra load due to styling (both in writing code and reviewing).

**I am labelling this PR as in need of a decision. I will wait for comments before doing more work.**

This PR does:

- [x] Add a pre-commit hooks for Black and `flake8`. The goal is to remove any churn on the development side. Just commit and it fixes the code for you, no questions asked.
- [x] black configuration is in `pyproject.toml` and `flake8` ones in `tox.ini`.
- [x] Make the max line length 88, Black's default.
- [x] Run Black on the whole code base.
- [ ] Add Black and `flake8` in CI (where?).
- [ ] Add isort (still not sure, maybe latter as this could break things with circular dependencies in some places).
- [ ] Remove `pycodestyle` from CI and `tox.ini`.
- [ ] Document Black use in the developer guides.

### Blame
We can mark the commit which applied Black to an exception list `.git-blame-ignore-revs`, see [here](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame), so that `git` does not see the commit and does not pollute the output of `git blame`.

This only works locally as GitHub does not support (yet) this functionality.

### Maths
There are maths concerns with operators. For instance these examples:

 https://github.com/tupui/scipy/blob/black_style/scipy/special/_basic.py#L1154 
```python
fac2 = (-1.0) ** (n + 1) * gamma(n + 1.0) * zeta(n + 1, x)
```

https://github.com/tupui/scipy/blob/black_style/scipy/optimize/tests/test_slsqp.py#L67
```python
dfdx = sign * (-2 * x + 2 * y + 2)
# vs
dfdx = sign*(-2*x + 2*y + 2)
```

My personal view on this: our current maths is not consistent on that, thus I prefer to pay this price (having more spacing) in favour of consistency. With Black, we remove all discussions around styling once and for all. Scikit-Learn did it (cc @thomasjpfan, @ogrisel), why not us.

I would also prefer to avoid having to fork Black to adjust such things. For `**`, there is an open issue/PR https://github.com/psf/black/issues/538 , so it could use some discussion to make it happen. Spacing around single operator is less likely to happen, but who knows... (cc @ambv)

### Skip Black
In case you really want to format something in a specific way, there is still the possibility to mark the code block, see the [doc](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#code-style).

### Merge conflicts
There are a few recipes out there to help fix merge conflicts with open PRs. We can link these in the doc.

https://github.com/scikit-learn/scikit-learn/issues/20301 https://github.com/spyder-ide/spyder/pull/15886

## References
Design choice, why the sad face https://lukasz.langa.pl/1d1a43c4-9c8a-4c5f-a366-7f22ce6a49fc/
Sklearn is now using Black  https://github.com/scikit-learn/scikit-learn/pull/18948
Django proposal for Black https://github.com/django/deps/blob/main/accepted/0008-black.rst

### CODE REVIEW

**Comment by rgommers** [comment]
> We can mark the commit which applied Black to an exception list `.git-blame-ignore-revs`, see [here](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame), so that `git` does not see the commit and does not pollute the output of `git blame`.

This would be easy to tests separately, and a useful thing to have in itself. There are some really old commits (from like 10 years ago) with "whitespace cleanup" (or some variant of that) in the title. How about doing a separate PR that adds  `.git-blame-ignore-revs` and skips those commits? Then we can very easily see that it works.

**Comment by ev-br** [comment]
-1 from me. Precisely because of the math formatting: e.g. `(2*a + 3*b)**2`. If black or some version of it can be made to preserve the binop priority, then we can reconsider IMO.

**Comment by prisae** [comment]
If black comes forward I would appreciate it to see the max line length remain at 79. Anyone in the world not working on a big fancy screen and/or working with terminals or traditional editors will be very thankful.

**Comment by perimosocordiae** [comment]
I see that a lot of the changes here are converting single-quoted strings to use double-quotes. Black now has an option to avoid that: --skip-string-normalization

**Comment by tupui** [comment]
> If black comes forward I would appreciate it to see the max line length remain at 79. Anyone in the world not working on a big fancy screen and/or working with terminals or traditional editors will be very thankful.

The thing with black is that it tends to take more vertical space. Hence, limiting a bit more the horizontal space will make it worse. I am not opposed to going back to 79, just that small screen size is not a bullet proof argument. Consider reading about this https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length and also the linked video is nice https://www.youtube.com/watch?v=wf-BqAjZb8M&t=260s

> I see that a lot of the changes here are converting single-quoted strings to use double-quotes. Black now has an option to avoid that: --skip-string-normalization

Do we want that? Wouldn't it break uniformity?

**Comment by rkern** [comment]
I'm personally -1 on `black`. I think it generally makes the code less readable, especially so for our math-heavy codebase, in exchange for mild improvements to writability. As we read code much more than write it, this is not a favorable exchange, to me.

**Comment by adeak** [comment]
As an outsider: do you have noticeable review overhead from bikeshedding style points? My understanding is that the main selling point of `black` is to prevent all such arguments, because it looks the way black wants it to look. Are such disagreements common(ish) in SciPy?

I guess there's also the overhead of not having to style new code correctly in the first place... but knowing myself I probably couldn't get myself to ignore style anyway, even if I knew that the code would be fed to `black` :)

**Comment by tupui** [comment]
> As an outsider: do you have noticeable review overhead from bikeshedding style points? 

IMO yes we do. At least the PR I reviewed, but I am a fairly new maintainer.

**Comment by rkern** [comment]
In my experience, there aren't many bikeshedding _arguments_, per se, ("I like it _this_ way!", "But I like it _that_ way!"). The main review overhead is just getting the style into compliance with the existing guidelines. `black`'s distinguishing feature is, as you say, more about resolving the former.

In general, I'd dearly love to have an autoformatting tool that only reformats code that breaks the style guidelines. That'd be truly useful and avoid the downsides of purely algorithmic reformatting.

**Comment by Carreau** [comment]
On IPython I use [darker](https://pypi.org/project/darker/) which only apply blacks formatting to newly added and modified lines (plus sometime the neighboring lines/block).

This and giving darker a base commit (in my case `60625f241f298b5039cb2debc365db38aa7bb522`) mean that all the code since then is black compliant, and there is no need for a `.git-blame-ignore-revs`, and rewriting whole files in a massive commit.

Because it uses a base commit instead of each commit independently , It's supposed to converge on all the code being black compliant at some point in the future (assuming each line will be close to a modification), or at least highly reduce the changes if we decide to switch to black later on.

It also has the option to apply isort and pyupgrade IIRC – which I'm not using though. 

As a maintainer, I'm personally a huge fan of only formating what has been touched since commit X. It means the blame and history is accurate both locally and on GitHub. I've tried [`.git-blame-ignore-revs` in IPython](https://github.com/ipython/ipython/blob/master/.git-blame-ignore-revs) but itis  an &lt;expletive&gt; pain.

--- 

Personally I also have it as my formatter for my editor, which mean that it also autoformat contributions I make to other projects – whether or not they black as an autoformatter

 - if they use black, then formatting only my change is the same as using black. 
 - if they don't it does not reformat their whole file for no reason.

**Comment by stefanv** [comment]
Along the lines of what @rkern suggests, a PEP-autofixer would be great.

`fixit` from Instagram seems to be close: https://fixit.readthedocs.io/en/latest/

(I found that in the list of tools that [gray](https://github.com/dizballanze/gray) builds upon. Other tools include [isort](https://timothycrosley.github.io/isort/) and [autoflake](https://github.com/myint/autoflake)---for removing unused variables).

**Comment by tupui** [comment]
Thank you everyone for the discussion and your inputs. From the conclusion here (and mostly the linked issue), I am closing this PR for now until we have a better support of mathematical equations from Black. Only then we could reconsider.
